### PR TITLE
Rem 191

### PR DESCRIPTION
**Pull request by FatimaSalami**

### CODE REVIEW

**Comment by NoraBitarSM** [comment]
https://github.com/GeorgeKalash/ArgusFE/assets/85101826/cedfb1c9-cabe-4dbe-861f-ddc84be0c579

File: src/repositories/SystemRepository.js [hunk 1]
```diff
@@ -146,7 +146,10 @@ export const SystemRepository = {
   FiscalYears: {
     qry: service + 'qryFY'
   },
-  ResourceControls: {
```
Comment (hadichahine) [inline src/repositories/SystemRepository.js:149]: Same to delete this?

File: src/components/Shared/Table.js [hunk 2]
```diff
@@ -318,7 +318,7 @@ const Table = ({
         const isWIP = row.wip === 2
 
         return (
-          <>
+          <Box sx={{ display: 'flex', width: '100%', justifyContent: 'center'}}>
```
Comment (hadichahine) [inline src/components/Shared/Table.js:321]: Do we need this?

File: src/components/Inputs/CustomComboBox.js [hunk 3]
```diff
@@ -138,6 +139,7 @@ const CustomComboBox = ({
         <TextField
           {...params}
           type={type}
+          onKeyDown={(e) => !editable && e.preventDefault()}
```
Comment (hadichahine) [inline src/components/Inputs/CustomComboBox.js:142]: Would this prevent us from clicking tab?

File: src/pages/global-authorization/forms/AccessLevelForm.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    const invalidate = useInvalidate({
+        endpointId: SystemRepository.ModuleClassRES.qry
+      })
+
+    const formik = useFormik({
+        initialValues,
+        enableReinitialize: true,
+        validateOnChange: true,
+        validationSchema: yup.object({
+          //accessLevel: yup.string().required('')
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/AccessLevelForm.js:43]: Do w need validation here? (to prevent submitting empty forms)

File: src/pages/global-authorization/forms/AccessLevelForm.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+          const updatedData = data.list.map(item => ({
+                moduleId:moduleId,
+                resourceId: item.key,
+                accessLevel: obj.accessLevel
+            })
+          )
+
+          updatedData.forEach(item => {
+            item.accessLevel != "" ? item.accessLevel = item.accessLevel : item.accessLevel = 0
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/AccessLevelForm.js:56]: Confusing. Should be `item.accessLevel = item.accessLevel || 0` in case we want to revert undefined, null, and empty string to 0.

File: src/pages/global-authorization/forms/AccessLevelForm.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+{
+
+    const [isLoading, setIsLoading] = useState(false)
+
+    const [initialValues, setInitialData] = useState({
+      accessLevel: '',
+      accessLevelName: ''
+   })
+
+    const { getRequest, postRequest } = useContext(RequestsContext)
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/AccessLevelForm.js:32]: Not using getRequest here right?

File: src/pages/global-authorization/forms/AccessLevelForm.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+          toast.success('Record Edited Successfully')
+          invalidate()
+        }
+      })
+
+      useEffect(() => {
+        ;(async function () {
+          try {
+
+            setIsLoading(true)
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/AccessLevelForm.js:72]: Why do we want loading here?

File: src/pages/global-authorization/forms/AccessLevelForm.js [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+          const updatedData = data.list.map(item => ({
+                moduleId:moduleId,
+                resourceId: item.key,
+                accessLevel: obj.accessLevel
+            })
+          )
+
+          updatedData.forEach(item => {
+            item.accessLevel != "" ? item.accessLevel = item.accessLevel : item.accessLevel = 0
+            postRequest({
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/AccessLevelForm.js:57]: Let's wait on postRequest to finish before we exit the function. Turn the wrapping function into async and add an await keyword here. It's necessary to let formik's onSubmit be an async function here. That way FormShell (later on) can ask formik if it's submitting to show a loading screen.

File: src/pages/global-authorization/forms/AccessLevelForm.js [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+          datasetId={DataSets.ACCESS_LEVEL}
+          name='accessLevel'
+          label={labels.accessLevel}
+          valueField='key'
+          displayField='value'
+          values={formik.values}
+          
+          maxAccess={maxAccess}
+          onChange={(event, newValue) => {
+            formik && formik.setFieldValue('accessLevel', newValue?.key)
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/AccessLevelForm.js:102]: No need to test if formik isn't null or undefined.

File: src/pages/global-authorization/forms/AccessLevelForm.js [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+          }}
+          error={formik.touched.accessLevel && Boolean(formik.errors.accessLevel)}
+          editable = {false}
+
+          //helperText={formik.touched.accessLevel && formik.errors.accessLevel}
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/AccessLevelForm.js:107]: Remove comments if unnecessary.

File: src/pages/global-authorization/index.js [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+  const [moduleIdSelection, setModuleIdSelection] = useState(10) //check new way
+
+  async function fetchGridData() {
+    return await getRequest({
+      extension: SystemRepository.ModuleClassRES.qry,
+      parameters: `_filter=&_moduleId=10`
+    })
+  }
+
+  async function fetchWithFilter({ filters }) {
```
Comment (hadichahine) [inline src/pages/global-authorization/index.js:33]: Both functions  appear similar. Let's combine them into one. Also stick to ?? instead of || here. ?? is a null check. || is a falsy check (false values)

File: src/pages/global-authorization/index.js [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    filter: {
+      endpointId: SystemRepository.ModuleClassRES.qry,
+      filterFn: fetchWithFilter
+    }
+  })
+
+  const onChange = value => {
+    if (value) {
+      filterBy('moduleId', value)
+      setModuleIdSelection(value)
```
Comment (hadichahine) [inline src/pages/global-authorization/index.js:61]: Why do we need this if it's inside filters?

**Review by hadichahine — CHANGES_REQUESTED** [review]
Reviewed most of the changes

File: src/pages/global-authorization/forms/FieldGlobalForm.js [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+      name: 'accessLevelCombo',
+      label: labels.accessLevel,
+      props: {
+        datasetId: DataSets.AU_RESOURCE_CONTROL_ACCESS_LEVEL,
+        valueField: 'key',
+        displayField: 'value',
+        mapping: [ { from: 'key', to: 'accessLevel' } ,
+          { from: 'value', to: 'accessLevelName' }]
+      },
+      async onChange({ row: { update, newRow } }) {
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/FieldGlobalForm.js:96]: I don't think we need an onChange here. The datagrid updates everything by itself. We used to use it when we didn't have mapping. It's used now in cash invoice to do numerical calculations whenever a number changes.

File: src/pages/global-authorization/forms/FieldGlobalForm.js [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+          const finalList = res.list.map(controlDetail => {
+            const n = {
+              controlId: controlDetail.id,
+              name: controlDetail.name,
+              accessLevel: null,
+              accessLevelName: null
+            };
+    
+            const matching = accessLevelRes.list.find(
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/FieldGlobalForm.js:130]: Sounds like confusing logic to me here. Can we get a brief explanation on why should we do such a mapping?

File: src/pages/global-authorization/forms/FieldGlobalForm.js [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+              y => n.controlId === y.controlId
+            );
+    
+            if (matching) {
+              n.accessLevel = matching.accessLevel;
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/FieldGlobalForm.js:135]: Let's go for naming other than n, x, y, etc.

File: src/pages/global-authorization/forms/FieldGlobalForm.js [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+    }
+  ]
+
+  useEffect(() => {
+    ;(async function () {
+      try {
+        if (resourceId) {
+          setIsLoading(true)
+
+          const res = await getRequest({
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/FieldGlobalForm.js:112]: Can we use useResourceQuery instead? Has isLoading out of the box.

File: src/pages/global-authorization/forms/ResourceGlobalForm.js [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+      })
+
+      useEffect(() => {
+        ;(async function () {
+          try {
+
+            if (resourceId) {
+              setIsLoading(true)
+
+              const res = await getRequest({
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/ResourceGlobalForm.js:71]: useResourceQuery instead of this?

**Review by hadichahine — CHANGES_REQUESTED** [review]
Make sure that all files are prettified. Let's also replace all references of `useFormik` with `useForm`.

File: src/components/Shared/Table.js (hunk 2 above)
Comment (FatimaSalami) [reply src/components/Shared/Table.js:321]: this set the icons (edit or delete) in center of column cell disregarding dimensions and if one of them is hidden or no.. Now you can notice that once one of them is hidden the position is not accurate.

File: src/components/Inputs/CustomComboBox.js (hunk 3 above)
Comment (FatimaSalami) [reply src/components/Inputs/CustomComboBox.js:142]: yes it affects this but after additional change it is fixed now

File: src/pages/global-authorization/forms/FieldGlobalForm.js (hunk 14 above)
Comment (FatimaSalami) [reply src/pages/global-authorization/forms/FieldGlobalForm.js:130]: here we fill the grid from 2 apis.. As if we have a class with 4 fields each one from a separate api so the first request we get the 2 fields we need and change their names to suit our grid binding and post.. Then we compare the id and get the rest of info from second request.

File: src/pages/global-authorization/index.js (hunk 12 above)
Comment (FatimaSalami) [reply src/pages/global-authorization/index.js:61]: i need to store value of combo selection  inside moduleIdSelection inorder to pass it to the forms.. Do you suggest another way?

File: src/pages/global-authorization/forms/FieldGlobalForm.js (hunk 16 above)
Comment (FatimaSalami) [reply src/pages/global-authorization/forms/FieldGlobalForm.js:112]: i guess here it is better to keep it query.. and what about has isloading? what do you mean?

File: src/pages/global-authorization/forms/ResourceGlobalForm.js (hunk 17 above)
Comment (FatimaSalami) [reply src/pages/global-authorization/forms/ResourceGlobalForm.js:71]: this is get not query, i have never realize using it for such a case..

File: src/pages/global-authorization/index.js (hunk 12 above)
Comment (FatimaSalami) [reply src/pages/global-authorization/index.js:61]: fixed

File: src/pages/global-authorization/forms/FieldGlobalForm.js [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+    maxAccess,
+    enableReinitialize: true,
+    validateOnChange: true,
+    initialValues: {
+      rows: [],
+      resourceId: resourceId,
+      resourceName: resourceName
+    },
+    onSubmit: async obj => {
+      console.log(detailsFormik.values.rows)
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/FieldGlobalForm.js:35]: Drop this log.

File: src/pages/global-authorization/forms/FieldGlobalForm.js [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+
+      const updatedRows = detailsFormik.values.rows
+        .filter(obj => obj.accessLevel != null && obj.accessLevel !== '')
+        .map(({ ...rest }) => ({
+          resourceId: resourceId,
+          ...rest
+        }))
+
+      console.log(detailsFormik.values.rows)
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/FieldGlobalForm.js:44]: Same here.

File: src/pages/global-authorization/forms/FieldGlobalForm.js [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+      rows: [],
+      resourceId: resourceId,
+      resourceName: resourceName
+    },
+    onSubmit: async obj => {
+      console.log(detailsFormik.values.rows)
+
+      const updatedRows = detailsFormik.values.rows
+        .filter(obj => obj.accessLevel != null && obj.accessLevel !== '')
+        .map(({ ...rest }) => ({
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/FieldGlobalForm.js:39]: no need to deconstruct into rest here.

File: src/pages/global-authorization/forms/ResourceGlobalForm.js [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+
+      toast.success('Record Edited Successfully')
+      invalidate()
+    }
+  })
+
+  useEffect(() => {
+    ;(async function () {
+      if (resourceId) {
+        setIsLoading(true)
```
Comment (hadichahine) [inline src/pages/global-authorization/forms/ResourceGlobalForm.js:59]: We're setting isLoading but I don't see any place loading is being referenced. We can drop it.

**Review by hadichahine — CHANGES_REQUESTED** [review]
